lib/tapi/tapi_cfg_sensor.c

```diff
@@ -0,0 +1,110 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright (C) OKTET Ltd. All rights reserved. */
+/** @file
+ * @brief Test API to get info about sensors.
+ *
+ * Implementation of API to get info about sensors.
+ */
+
+#define TE_LGR_USER      "Conf sensors TAPI"
+
+#include "te_config.h"
+
+#include <stdio.h>
+#ifdef STDC_HEADERS
+#include <stdlib.h>
+#include <string.h>
+#endif
+#ifdef HAVE_MATH_H
+#include <math.h>
+#endif
+
+#include "te_defs.h"
+#include "te_errno.h"
+#include "te_str.h"
+#include "te_string.h"
+
+#include "conf_api.h"
+#include "tapi_cfg.h"
+#include "tapi_cfg_base.h"
+#include "tapi_cfg_sensor.h"
+
+te_errno
+tapi_cfg_sensor_get(const char *tree_path, const char *error_msg,
+                    const char *ta, const char *sensor_name,
+                    const char *data_id_str, const char *threshold_id,
+                    double *value)
+{
+    char *value_str;
+    te_errno rc;
+
+    rc = cfg_get_instance_string_fmt(&value_str, tree_path, ta,
+                                     sensor_name, data_id_str,
+                                     threshold_id);
+    if (rc != 0)
+    {
+        ERROR(error_msg, sensor_name, data_id_str, threshold_id, rc);
+        return rc;
+    }
+
+    if (*value_str == '\0')
+    {
+        /* Value is not set */
+        *value = NAN;
+        return 0;
+    }
+    else
+    {
+        return te_strtod(value_str, value);
+    }
+}
+
+te_errno
+tapi_cfg_sensor_get_value(const char *ta, const char *sensor_name,
+                          const char *data_id_str, double *value)
+{
+    static const char path[] = "/agent:%s/hardware:/sensor:%s/data:%s"
+                                 "/value:%s";
+    static const char err_msg[] = "Failed to get value property of sensor"
+                                  " %s, data_id %s%s: error %r";
```

Comment:
That's unfortunate, because it means format strings become non-constant (mind you: in C unlike C++ `const` values are not actually treated as compile-time constants)

Reply:
So I see that you've changed `static const char` to `static char`. I sincerely do not understand what you could possibly meant, perhaps you misunderstood me. What I mean is this. In
```
static const char path[]  = "/agent:%s/hardware:/sensor:%s/data:%s/value:%s";
return tapi_cfg_sensor_get(path, err_msg, ta, sensor_name,
                               data_id_str, "", value);
```
`path` is not a compile-time constant, so the compiler would not in general be able to statically check the correctness of a `path` format. The presence or absence of `const` modifier does not impact this at all.

However, looking more closely, I see that the whole approach is flawed: one cannot handle format strings like you do safely.

What you probably should do instead is to factor out the unchanging part into a format string and pass the rest as plain strings. That is:
```
static te_errno
tapi_cfg_sensor_get(const char *ta, const char *sensor_name,
                                    const char *data_id_str, const char *threshold_id, const char *param_name, 
                                    const char *param_id, double *value)
{
    char *value_str;
    te_errno rc;

    rc = cfg_get_instance_string_fmt(&value_str, 
                                     "/agent:%s/hardware/sensor:%s/data:%s%s%s/%s:%s", ta,
                                     sensor_name, data_id_str,
                                     threshold_id == NULL ? "" : "/threshold:",
                                     threshold_id == NULL ? "" : threshold_id,
                                     param_name, param_value);
    if (rc != 0)
    {
        ERROR("Failed to get %s of sensor %s%s: %r", param_name, sensor_name, data_id_str, rc);
        return rc;
    }
    ...
```

And then in individual functions:
```
te_errno
tapi_cfg_sensor_get_value(const char *ta, const char *sensor_name,
                                                const char *data_id_str, double *value)
{
    return tapi_cfg_sensor_get(ta, sensor_name,  data_id_str, NULL, "value", value);
}
```